### PR TITLE
Optimize title search by removing redundant cycles

### PR DESCRIPTION
**Pull request by Jaifroid**
This is an optimization of title search based on my work introducing wildcard search in https://github.com/kiwix/kiwix-js-windows/issues/165. I discovered that for each prefix variant (by case) we were potentially searching and reading between 25 and 50 more dirEntries than was necessary. For many variants, only one or two results are found very quickly, and no further results in the title index are relevant because they are beyond the alphabetical range of the prefix. For titles with many case variants, this was producing a lot of churn.

This works by stopping a variant search as soon as the prefix variant no longer matches the dirEntry's title, rather than by setting an artificial search window. It also stops searching as soon as `params.maxSearchResultsSize` results are found. This produces very tight searches with no redundancy.

There is a diagnostic block which produces a readout in console debug showing the number of titles scanned and the number of results found per prefix variant. If you search for something with lots of variants like "unesco world heritage site" (in full English Wikipedia) you will see that titles scanned are never greater than results found + 1, which is the maximum efficiency possible. Results are returned pretty instantly.

### CODE REVIEW

**Comment by Jaifroid** [comment]
Converting to draft while I check test failure.

**Comment by Jaifroid** [comment]
I forgot to adapt tests to altered title search API.

File: www/js/lib/zimArchive.js
Comment (Jaifroid) [inline www/js/lib/zimArchive.js:290]: Three lines above to be removed before merge.

**Comment by Jaifroid** [comment]
No passing tests, can be reviewed.

**Review by mossroy — APPROVED** [review]
Quickly tested on latest Chromium and my Firefox OS device : I saw no regression
On my Firefox OS device, when there are a lot of variants (I hate this word nowadays ;-) ), this version is twice as fast.
If I only type a few letters in the search form, or even 2 words, I don't see any speed difference (which looks logical).
I did not look deeply at the algorithm change, but you seem to master the topic.

File: www/js/lib/zimArchive.js
Comment (Jaifroid) [reply www/js/lib/zimArchive.js:265]: Ah yes, this was overzealousness. I'm not really sure it's necessary to define a whole new type for a callback function, but it is properly defined in the module. Do you think it's useful that this module defines a new type? I'm happy to revert it.

**Comment by Jaifroid** [comment]
> On my Firefox OS device, when there are a lot of variants (I hate this word nowadays ;-) ), this version is twice as fast.

😉 I was thinking the same about "variants"....
I'm glad you notice a tangible benefit on a slower device!

File: www/js/lib/zimArchive.js
Comment (mossroy) [reply www/js/lib/zimArchive.js:265]: Depending on the IDE you use, it can allow some autocompletion and controls.
It's clearly not compulsory, but can ease development in some cases

Comment (Jaifroid) [reply www/js/lib/zimArchive.js:265]: OK, will revert this change.

**Comment by Jaifroid** [comment]
Reminder to self: test thoroughly on no-namespace ZIMs before merge (works OK on one quick test).

**Comment by Jaifroid** [comment]
@mossroy I added some usage notes explaining how title search works in the latest commit. Please check the wording is OK for you.

I would also recommend we do the following now that title search has eliminated most redundancy:

* Increase slider range to 100 (current max 50), for users who want a real in-depth search - there's no reason to limit it to 50
* Increase initial preset from 25 to 30
* Increase the number of words allowed for the "full" case variant search type from 5 to 6. The algorithm now handles easily and pretty fast 3^6 case combinations (729) compared to our current max of 3^5 (243)..

Each of these can be enabled by changing a few hard-coded values. But I realize these are all extras for edge cases or advanced users, so I leave it to you to decide!

**Review by mossroy — APPROVED** [review]
The wording looks good to me. It's a good idea to add this explanation.

Regarding the hard-coded values, I don't have a strong opinion. But we need to make sure we don't put a negative impact for most users, to be able to handle edge cases.

- Allowing the slider range to reach 100 instead of 50 should not harm
- I think 25 results is already more than enough for most users
- OK why not. 5 or 6 words is certainly already an edge case

File: www/index.html
Comment (Jaifroid) [reply www/index.html:202]: Yes, but maybe "full text search of the contents of articles and titles" just to be ultra-clear.

Comment (mossroy) [reply www/index.html:202]: Sure

**Comment by Jaifroid** [comment]
* Allowing the slider range to reach 100 instead of 50 should not harm - **done**
* I think 25 results is already more than enough for most users - **OK (left at 25)** 
* OK why not. 5 or 6 words is certainly already an edge case - **increased to 300 combination, allows 5 words**

I actually made a mistake in comment above. We currently search multiple case combination for up to **4** words. I've increased the maximum to **5** words. I tried 6, and there is an appreciable delay, so I think 5 is the maximum that still produces fast performance with the new title search algorithm in this PR.

**Comment by Jaifroid** [comment]
I just need to do some more tests on no-namespace ZIMs that we have (and that are actually working). I'll record results here, and then if all is satisfactory, I'll merge.

**Comment by Jaifroid** [comment]
I've now tested successfully on the no-ns ZIMs:

* `wikispecies_en_all_maxi_2020-06_nons.zim`
* `wikipedia_en_ray_charles_maxi_2021-03_nons.zim`
* `3dprinting.stackexchange.com_en_all_2018-03_nons.zim`
* `beer_stackexchange_com_2021-07.zim` (after commenting out the line that produces the unhandled exception)

I only tested these on Edge Chromium, because it was a test of the algorithm, not of browser support.

So I think this is now ready for squash/merge.

**Comment by mossroy** [comment]
On my side, you can squash/merge

**Pull request merged**
